src/Socket/newsletter.ts

```diff
@@ -0,0 +1,269 @@
+import type { NewsletterCreateResponse, WAMediaUpload } from '../Types'
+import { NewsletterMetadata, NewsletterUpdate, QueryIds, XWAPaths } from '../Types'
+import { generateProfilePicture } from '../Utils/messages-media'
+import { getBinaryNodeChild, S_WHATSAPP_NET } from '../WABinary'
+import { GroupsSocket } from './groups'
+
+const parseNewsletterCreateResponse = (response: NewsletterCreateResponse): NewsletterMetadata => {
+	const { id, thread_metadata: thread, viewer_metadata: viewer } = response
+	return {
+		id: id,
+		owner: undefined,
+		name: thread.name.text,
+		creation_time: parseInt(thread.creation_time, 10),
+		description: thread.description.text,
+		invite: thread.invite,
+		subscribers: parseInt(thread.subscribers_count, 10),
+		verification: thread.verification,
+		picture: {
+			id: thread.picture.id,
+			directPath: thread.picture.direct_path
+		},
+		mute_state: viewer.mute
+	}
+}
+
+const parseNewsletterMetadata = (result: unknown): NewsletterMetadata | null => {
+	if (typeof result !== 'object' || result === null) {
+		return null
+	}
+
+	if ('id' in result && typeof result.id === 'string') {
+		return result as NewsletterMetadata
+	}
+
+	if ('result' in result && typeof result.result === 'object' && result.result !== null && 'id' in result.result) {
+		return result.result as NewsletterMetadata
+	}
+
+	return null
+}
+
+export const makeNewsletterSocket = (sock: GroupsSocket) => {
+	const { query, generateMessageTag } = sock
+	const newsletterWMexQuery = async (variables: Record<string, unknown>, queryId: string) => {
+		const result = await query({
+			tag: 'iq',
+			attrs: {
+				id: generateMessageTag(),
+				type: 'get',
+				to: S_WHATSAPP_NET,
+				xmlns: 'w:mex'
+			},
+			content: [
+				{
+					tag: 'query',
+					attrs: { query_id: queryId },
+					content: Buffer.from(JSON.stringify({ variables }), 'utf-8')
+				}
+			]
+		})
+		return result
+	}
+
+	const executeWMexQuery = async <T>(
+		variables: Record<string, unknown>,
+		queryId: string,
+		dataPath: string
+	): Promise<T> => {
+		const result = await newsletterWMexQuery(variables, queryId)
+		const child = getBinaryNodeChild(result, 'result')
+		if (child?.content) {
+			const data = JSON.parse(child.content.toString())
+
+			if (data.errors && data.errors.length > 0) {
+				const errorMessages = data.errors.map((err: Error) => err.message || 'Unknown error').join(', ')
+				const firstError = data.errors[0]
+				const errorCode = firstError.extensions?.error_code || 400
+				throw new Boom(`GraphQL server error: ${errorMessages}`, { statusCode: errorCode, data: firstError })
+			}
+
+			const response = dataPath ? data?.data?.[dataPath] : data?.data
+			if (typeof response !== 'undefined') {
+				return response as T
+			}
+		}
+
+		const action = dataPath?.replace('xwa2_newsletter_', '').replace(/_/g, ' ')
+		throw new Boom(`Failed to ${action}, unexpected response structure.`, { statusCode: 400, data: result })
+	}
```

Comment:
Can we move all the mex stuff to another file or something? For consistency, as mex is used in other queries as well.

Reply:
Done